### PR TITLE
CONTRIB-7413: New feature: choose between Flash/HTML5 Web Client in each room

### PR DESCRIPTION
**Pull request by rabser**
As per your previous request, i did this pull request for letting you evaluate my changes for coding this new feature. I tried to follow your coding philosophy, so most of the changes to the code were done for a "correct" integration of the new setting, enabling/disabling it and it's default.
The semantic section is around the join API, were i added the joinViaHTML5 with value true if requested a define time of each single room.
The BBB server should be configured to serve the HTML5 client, but without forcing it. If the HTML5client is not installed, this option it's simply ignored.

### CODE REVIEW

File: version.php [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 defined('MOODLE_INTERNAL') || die;
 
-$plugin->version = 2017101014;
+$plugin->version = 2017101015;
```
Comment (jfederico) [inline version.php]: Let's dont bump up the version just yet. I usually do that right before release.

File: mod_form.php [hunk 2]
```diff
@@ -69,6 +69,7 @@ public function definition() {
         $context = context_course::instance($course->id);
         // UI configuration options.
         $cfg = \mod_bigbluebuttonbn\locallib\config::get_options();
+error_log("RABSER DEBUG ".print_r($cfg,true));
```
Comment (jfederico) [inline mod_form.php]: Debugging code has to be removed. This is going to clean a few errors that are making travis to fail.

File: locallib.php [hunk 3]
```diff
@@ -87,12 +87,17 @@
  *
  * @return string
  */
-function bigbluebuttonbn_get_join_url($meetingid, $username, $pw, $logouturl, $configtoken = null, $userid = null) {
+function bigbluebuttonbn_get_join_url($meetingid, $username, $pw, $logouturl, $configtoken = null, $userid = null, $clienttype=0) {
```
Comment (jfederico) [inline locallib.php]: For the default value and the comparisions, I would use a constant to make it more explicit.

**Comment by jfederico** [comment]
Thanks @rabser , it looks good over all. There are some minor things to fix, but this comes in a good timing. I am working on Moodle and I will for the next few weekes.

File: version.php (hunk 1 above)
Comment (jfederico) [reply version.php]: No, it is fine. The version tag has to be updated anyways. It is just that, as the code is going directly to v2.2-stable, there may be epole deploying from there and not from the moodle plugin database. Besides, the other tag has also be updated to 2.2.7.

Keep in mind that this is for the pull request only, once released both will be updated anyways.

And I can revert it directly myself before merging, but it is usually left to the contributor to clean up the code. I contribute with some other open source projects myself, I guess I am just getting used to it :)

**Comment by jfederico** [comment]
My concern is that travis is still not passing
https://travis-ci.org/blindsidenetworks/moodle-mod_bigbluebuttonbn/builds/437166743?utm_source=github_status&utm_medium=notification

But I don't have time to check it now. If you can do it that would be great, otherwise I will need to find some time in a couple of weeks.

**Comment by jfederico** [comment]
BTW, these are still issues in your code
```
FILE: /home/travis/build/moodle/mod/bigbluebuttonbn/mod_form.php
--------------------------------------------------------------------------------------------------
FOUND 9 ERRORS AND 2 WARNINGS AFFECTING 5 LINES
--------------------------------------------------------------------------------------------------
 395 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
 396 | ERROR   | [ ] Variable "clienttype_select_choices" must not contain underscores.
 396 | ERROR   | [ ] Line exceeds maximum limit of 180 characters; contains 247 characters
 400 | ERROR   | [ ] Commas (,) must be followed by white space.
 400 | ERROR   | [ ] Variable "clienttype_select_choices" must not contain underscores.
 402 | ERROR   | [ ] Commas (,) must be followed by white space.
 402 | WARNING | [ ] Line exceeds 132 characters; contains 147 characters
 403 | ERROR   | [x] Spaces must be used to indent lines; tabs are not allowed
 403 | ERROR   | [ ] Tab found within whitespace
 403 | ERROR   | [x] Line indented incorrectly; expected 8 spaces, found 1
 403 | ERROR   | [x] Closing brace indented incorrectly; expected 8 spaces, found 1
--------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------------------------
FILE: /home/travis/build/moodle/mod/bigbluebuttonbn/locallib.php
---------------------------------------------------------------------------------------------------
FOUND 11 ERRORS AND 4 WARNINGS AFFECTING 8 LINES
---------------------------------------------------------------------------------------------------
   94 | WARNING | [ ] Line exceeds 132 characters; contains 160 characters
  100 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
  102 | ERROR   | [x] Spaces must be used to indent lines; tabs are not allowed
  102 | ERROR   | [ ] Tab found within whitespace
  102 | ERROR   | [x] Line indented incorrectly; expected at least 8 spaces, found 5
 2609 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
 2610 | ERROR   | [ ] Variable "clienttype_default" must not contain underscores.
 2612 | ERROR   | [x] Spaces must be used to indent lines; tabs are not allowed
 2612 | ERROR   | [ ] Tab found within whitespace
 2612 | WARNING | [ ] Inline comments must end in full-stops, exclamation marks, or question marks
 2612 | ERROR   | [x] Line indented incorrectly; expected at least 8 spaces, found 1
 2613 | ERROR   | [ ] Variable "clienttype_select_choices" must not contain underscores.
 2613 | ERROR   | [ ] Line exceeds maximum limit of 180 characters; contains 243 characters
 2616 | ERROR   | [ ] Variable "clienttype_default" must not contain underscores.
 2616 | ERROR   | [ ] Variable "clienttype_select_choices" must not contain underscores.
---------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
---------------------------------------------------------------------------------------------------
```
But the big one is Behat, that should not be affected. The fact is that it is.

**Comment by rabser** [comment]
thanks for your clarifications, i'll try to fix all these errors.

**Comment by rabser** [comment]
It seems all ok, now. Thanks again for your patience.

**Comment by rabser** [comment]
it should be enough to check if the bbb_html5 client is running, it's a simple api that return success in json format. Thanks for your corrections, i forgot some things like the backup one.

**Comment by jfederico** [comment]
Yeah, that is what I did. I am validating against html5client/check, but only the HEAD. One more thing is that the validations of settings were not working properly.

And I restored the clienttype_enabled you added (which is working in the 4 cases it should). As we don't want to have this feature enabled for everybody, the  clienttype_enabled has to be explicitly set to 1 in config.php

You'll see the updates. It would be nice if you can give it a try, and also if you can run codechecker and phpdocs just to be sure. I didn;t have time to do that.

It is looking pretty cool actually! :+1: 
Thanks again for pushing this forward

**Pull request merged**